packages/react-components/react-tree/src/components/TreeItemLayout/useTreeItemLayout.tsx

```diff
@@ -60,14 +116,39 @@ export const useTreeItemLayout_unstable = (
   const arrowNavigationProps = useArrowNavigationGroup({ circular: true, axis: 'horizontal' });
   const actions = isActionsVisible
     ? slot.optional(actionsShorthand, {
-        defaultProps: { ...arrowNavigationProps, role: 'toolbar' } as ExtractSlotProps<TreeItemLayoutSlots['actions']>,
+        defaultProps: { ...arrowNavigationProps, role: 'toolbar' },
         elementType: 'div',
       })
     : undefined;
-  const actionsRefs = useMergedRefs(actions?.ref, actionsRef);
+  const actionsRefs = useMergedRefs(actions?.ref, actionsRef, actionsRefInternal);
+  const handleActionsBlur = useEventCallback((event: React.FocusEvent<HTMLDivElement>) => {
+    if (isResolvedShorthand(actionsShorthand)) {
+      actionsShorthand.onBlur?.(event);
+    }
+    setIsActionsVisibleInternal(elementContains(event.currentTarget, event.relatedTarget as Node));
+  });
   if (actions) {
     actions.ref = actionsRefs;
+    actions.onBlur = handleActionsBlur;
   }
+
+  React.useEffect(() => {
+    if (treeItemRef.current && actionsShorthand && isActionsVisibleExternal === undefined) {
```

Comment:
This condition is broken, you rely on presence of `actionsShorthand`, but you don't tracking `actionsShorthand` in dependencies. While you have `actions` in dependencies...

I propose to do something like that:
- we have clear dependencies
- we still explicit, but don't declare useless callbacks with `useEventCallback()`

```tsx
  const hasActions = Boolean(actionsShorthand);

  React.useEffect(() => {
    if (treeItemRef.current && hasActions && typeof isActionsVisibleFromProps === 'undefined') {
      const treeItemElement = treeItemRef.current;

      const handleMouseOver = setActionsVisibleIfNotFromSubtree;
      const handleMouseOut = setActionsInvisibleIfNotFromSubtree;

      const handleFocus = setActionsVisibleIfNotFromSubtree;
      const handleBlur = setActionsInvisibleIfNotFromSubtree;

      treeItemElement.addEventListener('mouseover', handleMouseOver);
      treeItemElement.addEventListener('mouseout', handleMouseOut);
      treeItemElement.addEventListener('focus', handleFocus);
      treeItemElement.addEventListener('blur', handleBlur);

      return () => {
        treeItemElement.removeEventListener('mouseover', handleMouseOver);
        treeItemElement.removeEventListener('mouseout', handleMouseOut);
        treeItemElement.removeEventListener('focus', handleFocus);
        treeItemElement.removeEventListener('blur', handleBlur);
      };
    }
  }, [
    hasActions,
    treeItemRef,
    isActionsVisibleFromProps,
    setActionsVisibleIfNotFromSubtree,
    setActionsInvisibleIfNotFromSubtree,
  ]);
```